src/render/painter.js

```diff
@@ -457,28 +458,39 @@ class Painter {
             this.renderLayer(this, sourceCache, layer, coords);
         }
 
+        this.depthRangeFor3D = [0, 1 - ((style.order.length + 2) * this.numSublayers * this.depthEpsilon)];
+
+        // Terrain depth offscreen render pass ==========================
+        // With terrain on, renders the depth buffer into a texture.
+        // This texture is used for occlusion testing (labels)
+        if (this.terrain && (this.style.hasLayerType('symbol') || this.style.hasLayerType('circle'))) {
```

Comment:
Its probably a very small performance impact, but `hasLayerType` will loop over all layers every frame, how about using a `Set` in `Style` to track used layer types instead.

Reply:
Updated for bookkeeping similarly to 3d layers in https://github.com/mapbox/mapbox-gl-js/pull/10258/commits/b732630d4465ce232a58c71932843833dc2885e9.